### PR TITLE
Improve tutorial.md's clarity and structure

### PR DESCRIPTION
**Pull request by jblang94**
This PR attempts to improve tutorial.md's clarity and structure.

Thank you for the PR! Contributors like you keep React awesome!

Please see the Contribution Guide for guidelines:

https://github.com/reactjs/reactjs.org/blob/master/CONTRIBUTING.md

If your PR references an existing issue, please add the issue number below:

#<Issue>

### CODE REVIEW

File: content/tutorial/tutorial.md [hunk 1]
```diff
@@ -978,26 +972,26 @@ to
...
+**It's strongly recommended that you assign proper keys whenever you build dynamic lists.** If you don't have an appropriate key, you may want to consider restructuring your data so that you do.
 
-If you don't specify any key, React will warn you and fall back to using the array index as a key – which is not the correct choice if you ever reorder elements in the list or add/remove items anywhere but the bottom of the list. Explicitly passing `key={i}` silences the warning but has the same problem so isn't recommended in most cases.
+If no key is specified, React will present a warning and use the array index as a key by default. Using the array index as a key is problematic when trying to re-order a list's items or inserting/removing list items. Explicitly passing `key={i}` silences the warning but has the same problems as array indices and is not recommended.
 
-Component keys don't need to be globally unique, only unique relative to the immediate siblings.
+Keys do not need to be globally unique. Keys only needs to be unique between components and their sibilings.
```
Comment (ngyikp) [inline content/tutorial/tutorial.md]: Minor typo: sibilings -> siblings

File: content/tutorial/tutorial.md [hunk 2]
```diff
@@ -494,29 +488,31 @@ var newPlayer = Object.assign({}, player, {score: 2});
...
+Detecting changes in mutable objects is complex because they are modified directly. This detection requires the mutable object to be compared to previous copies of itself and the entire object tree to be traversed.
 
-Determining how an immutable object has changed is considerably easier. If the object being referenced is different from before, then the object has changed. That's it.
+Detecting changes in immutable objects is considerably easier. If the immutable object that is being referenced is different than the previous one, then the object has changed.
 
 #### Determining When to Re-render in React
 
-The biggest benefit of immutability in React comes when you build simple _pure components_. Since immutable data can more easily determine if changes have been made, it also helps to determine when a component requires being re-rendered.
+The main benefit of immutability is that it helps you build _pure components_ in React. Immutable data can easily determine if changes have been made which helps to determine when a component requires re-renderering.
```
Comment (ngyikp) [inline content/tutorial/tutorial.md]: re-renderering -> re-rendering

File: content/tutorial/tutorial.md (hunk 1 above)
Comment (jblang94) [reply content/tutorial/tutorial.md]: Thanks, fixed in new commit

File: content/tutorial/tutorial.md (hunk 2 above)
Comment (jblang94) [reply content/tutorial/tutorial.md]: Thanks, fixed in new commit

**Comment by alexkrolick** [comment]
These look like good improvements at first skim, will have to get some more eyes on it before merging though. I think the core team is pretty busy with the 16.3 release this week but hopefully someone can take a look soon.

Thanks!

File: content/tutorial/tutorial.md [hunk 3]
```diff
@@ -85,12 +83,12 @@ We recommend following [these instructions](http://babeljs.io/docs/editors) to c
...
 ### Help, I'm Stuck!
 
-If you get stuck, check out the [community support resources](/community/support.html). In particular, [Reactiflux chat](/community/support.html#reactiflux-chat) is a great way to get quick help. If you don't get a good answer anywhere, please file an issue, and we'll help you out.
-
-With this out of the way, let's get started!
+If you get stuck, check out the [community support resources](/community/support.html). In particular, [Reactiflux Chat](/community/support.html#reactiflux-chat) is a great way to get help quickly. If you don't receive an answer, or if you remain stuck, please file an issue, and we'll help you out.
 
 ## Overview
 
+Now that you're setup, let's get an overview of React!
```
Comment (alexkrolick) [inline content/tutorial/tutorial.md]: nit: set up

**Comment by gaearon** [comment]
This is great work, thank you. I think these edits make everything more explicit and I personally found it easier to follow.

There were some minor things I disagreed with (e.g. I think there are a few slightly whimsical sentences that I think are worth keeping in their current form). There were also some minor technical inaccuracies.

I tried to strike a balance and keep the tone of your edits in the ones I made on top, but please let me know if you have ideas on rephrasing on top. I admit some of my writing is a bit clumsy, but I hope you can see what I was trying to do.

Overall major props and thanks

**Pull request merged**